### PR TITLE
gh-408 Refactor and port Analytics Dashboard tab

### PR DESCRIPTION
**Pull request by ghillert**
* Add `app-counter-header` directive
* Add Bar Chart

Resolves https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/408

### CODE REVIEW

File: ui/src/app/analytics/analytics.service.ts [hunk 1]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+    if (index) {
+      this.dashboardItems.splice(index, 0, dashboardItem);
+    } else {
+      this.dashboardItems.push(dashboardItem);
+    }
+    console.log('>>>', this.dashboardItems);
+    return dashboardItem;
+  }
+
+  removeDashboardItem(index: number): void {
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: JSDoc

File: ui/src/app/analytics/analytics.service.ts [hunk 2]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+    if (!this.dashboardItems || this.dashboardItems.length === 0) {
+      return;
+    }
+    this.dashboardItems.splice(index, 1);
+  }
+
+  getAllDashboardItems(): Observable<DashboardItem> {
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: JSDoc

File: ui/src/app/analytics/analytics.service.ts [hunk 3]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+    this.addNewDashboardItem();
+    return Observable.from(this.dashboardItems);
+  }
+
+  getStreamsForMetricType(metricType: MetricType) {
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: JSDoc

File: ui/src/app/analytics/analytics.service.ts [hunk 4]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+      return this.getAllCounters();
+    }
+  }
+
+  resetDashboard() {
+    this.dashboardItems.length = 0;
+    this.addNewDashboardItem();
+  }
+
+  //
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: Can be removed

File: ui/src/app/analytics/analytics.service.ts [hunk 5]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+
+    /**
+   * Starts the polling process for a single counters. Method
+   * will check if the poller is already running and will
+   * start the poller only if the poller is undefined or
+   * stopped. The subscription is store on the {@link DashboardItem}.
+   */
+  public startPollingForSingleDashboardItem(dashboardItem: DashboardItem) {
+    console.log(dashboardItem);
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: console.log can be removed.

File: ui/src/app/analytics/analytics.service.ts [hunk 6]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+  }
+
+  /**
+   * Restarts the polling process for counters if the poller
+   * is running and is defined. Will stop the poller first and restart
+   * the counter with the {@link DashboardItem}s refresh rate.
+   *
+   * Will NOT restart the poller if the refresh rate is zero.
+   */
+  public restartPollingOfSingleDashboarItem(dashboardItem: DashboardItem) {
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: Dashboard in function name is misspelled.

File: ui/src/app/analytics/analytics.service.ts [hunk 7]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+            this.toastyService.error(error);
+          });
+    }
+  }
+
+  /**
+   * Stops the polling process for counters if the poller
+   * is running and is defined.
+   */
+  public stopPollingOfSingleDashboarItem(dashboardItem: DashboardItem) {
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: Dashboard is misspelled

File: ui/src/app/analytics/analytics.service.ts [hunk 8]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+      this.startPollingForSingleDashboardItem(dashboardItem);
+    }
+  }
+
+  /**
+   * Retrieves all counters. Will take pagination into account.
+   *
+   * @param detailed If true will request additional counter values from the REST endpoint
+   */
+  public getSingleCounter(counterName: string): Observable<Counter> {
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: Should be private.

File: ui/src/app/analytics/analytics.service.ts [hunk 9]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+    this.stopPollingOfSingleDashboarItem(dashboardItem);
+    if (dashboardItem.refreshRate > 0 ) {
+      this.startPollingForSingleDashboardItem(dashboardItem);
+    }
+  }
+
+  /**
+   * Retrieves all counters. Will take pagination into account.
+   *
+   * @param detailed If true will request additional counter values from the REST endpoint
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: JSDoc copy paste hiccup.

File: ui/src/app/analytics/charts/bar-chart/bar-chart.component.ts [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+          || this.barSizeCopy !== this.barSize) {
+      this.chartDataCopy = this.chartData.slice();
+      this.widthCopy = this.width;
+      this.barSizeCopy = this.barSize;
+      this.createChart();
+    }
+  }
+
+  private isArraySame(array1: number[], array2: number[]): boolean {
+    if ((!array1 && array2) || (array2 && !array2)) {
```
Comment (cppwfs) [inline ui/src/app/analytics/charts/bar-chart/bar-chart.component.ts]: Did you mean it this way: if ((!array1 && array2) || (array`1` && !array2))

File: ui/src/app/analytics/charts/bar-chart/bar-chart.component.ts [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+      if (array1[i] !== array2[i]) {
+          return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Re-render the chart, in case the window is resized in
+   * order to maintain dimentsions.
```
Comment (cppwfs) [inline ui/src/app/analytics/charts/bar-chart/bar-chart.component.ts]: dimensions is misspelled

File: ui/src/app/analytics/charts/bar-chart/bar-chart.component.ts [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    if (!this.chartData) {
+      return;
+    } else {
+      console.log(`Render Chart - width=${widthToUse}, height=${heightToUse}, data: ${this.chartData} - ` + Array.isArray(this.chartData));
+    }
+
+    const reversedChartData = this.chartData.slice();
+    reversedChartData.reverse();
+
+    const margin = {top: 20, right: 10, bottom: 20, left: 10};
```
Comment (cppwfs) [inline ui/src/app/analytics/charts/bar-chart/bar-chart.component.ts]: margin is not used.  can be removed.

File: ui/src/app/analytics/components/counter-header.component.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import { Component, Input } from '@angular/core';
```
Comment (cppwfs) [inline ui/src/app/analytics/components/counter-header.component.ts]: Need to add spec with a toBeTruthy test.

File: ui/src/app/analytics/counters/counters.component.ts [hunk 14]
```diff
@@ -4,7 +4,7 @@ import { Subscription } from 'rxjs/Subscription';
 import { AnalyticsService } from '../analytics.service';
```
Comment (cppwfs) [inline ui/src/app/analytics/counters/counters.component.ts]: subscription, page and modaldirective can be removed from imports.

File: ui/src/app/analytics/dashboard/dashboard.component.ts [hunk 15]
```diff
@@ -2,6 +2,11 @@ import { Component, OnInit } from '@angular/core';
 import { Router } from '@angular/router';
 import { Subscription } from 'rxjs/Subscription';
 
+import { AnalyticsService } from '../analytics.service';
```
Comment (cppwfs) [inline ui/src/app/analytics/dashboard/dashboard.component.ts]: Router import can be removed.

Comment (cppwfs) [inline ui/src/app/analytics/dashboard/dashboard.component.ts]: Dashboard component needs a toBeTruthy spec.

File: ui/src/app/analytics/dashboard/dashboard.component.ts [hunk 16]
```diff
@@ -14,7 +19,79 @@ import { Subscription } from 'rxjs/Subscription';
...
+  /**
+   * Adds a new {@link DashboardItem} at the specified index
+   * @param index Insertion index
+   */
+  addNewDashboardItem(index: number) {
+    console.log('Adding new DashboardItem after index ' + index);
+    this.analyticsService.addNewDashboardItem(index + 1);
+  }
+
+  moveUpItem(dashboardItem: DashboardItem, currentIndex: number) {
```
Comment (cppwfs) [inline ui/src/app/analytics/dashboard/dashboard.component.ts]: JsDoc

File: ui/src/app/analytics/dashboard/dashboard.component.ts [hunk 17]
```diff
@@ -14,7 +19,79 @@ import { Subscription } from 'rxjs/Subscription';
...
+    this.dashboardItems.splice(currentIndex - 1, 0, this.dashboardItems.splice(currentIndex, 1)[0]);
+  }
+  moveDownItem(dashboardItem: DashboardItem, currentIndex: number) {
```
Comment (cppwfs) [inline ui/src/app/analytics/dashboard/dashboard.component.ts]: JsDoc

File: ui/src/app/analytics/dashboard/dashboard.component.ts [hunk 18]
```diff
@@ -14,7 +19,79 @@ import { Subscription } from 'rxjs/Subscription';
...
+    this.dashboardItems.splice(currentIndex + 1, 0, this.dashboardItems.splice(currentIndex, 1)[0]);
+  }
+
+  resetDashboard() {
```
Comment (cppwfs) [inline ui/src/app/analytics/dashboard/dashboard.component.ts]: JsDoc

File: ui/src/app/analytics/dashboard/dashboard.component.ts [hunk 19]
```diff
@@ -14,7 +19,79 @@ import { Subscription } from 'rxjs/Subscription';
...
+    this.analyticsService.resetDashboard();
+  }
+
+  onMetricTypeChange(metricType: MetricType) {
+    console.log('Selected Metric Type:', metricType);
+    this.analyticsService.getStreamsForMetricType(metricType).subscribe(result => {
+      console.log(result);
```
Comment (cppwfs) [inline ui/src/app/analytics/dashboard/dashboard.component.ts]: remove console.log

File: ui/src/app/analytics/dashboard/dashboard.component.ts [hunk 20]
```diff
@@ -14,7 +19,79 @@ import { Subscription } from 'rxjs/Subscription';
...
+  }
+  moveDownItem(dashboardItem: DashboardItem, currentIndex: number) {
+    this.dashboardItems.splice(currentIndex + 1, 0, this.dashboardItems.splice(currentIndex, 1)[0]);
+  }
+
+  resetDashboard() {
+    this.analyticsService.resetDashboard();
+  }
+
+  onMetricTypeChange(metricType: MetricType) {
```
Comment (cppwfs) [inline ui/src/app/analytics/dashboard/dashboard.component.ts]: JsDoc

File: ui/src/app/analytics/dashboard/dashboard.component.ts [hunk 21]
```diff
@@ -14,7 +19,79 @@ import { Subscription } from 'rxjs/Subscription';
...
+    console.log('Selected Metric Type:', metricType);
+    this.analyticsService.getStreamsForMetricType(metricType).subscribe(result => {
+      console.log(result);
+      this.counters = result.items;
+    });
+  }
+
+  onCounterNameChange(dashBoardItem: DashboardItem) {
```
Comment (cppwfs) [inline ui/src/app/analytics/dashboard/dashboard.component.ts]: add JSDOcs to all public methods.

File: ui/src/app/analytics/model/dashboard-item.model.ts [hunk 22]
```diff
@@ -0,0 +1,42 @@
+import { Serializable } from '../../shared/model';
```
Comment (cppwfs) [inline ui/src/app/analytics/model/dashboard-item.model.ts]: Needs a spec.

File: ui/src/app/analytics/model/dashboard-item.model.ts [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+  /**
+   * For a given JSON data object, this method
+   * will populate the corresponding Counter object with
+   * the provided properties.
+   *
+   * @param input JSON input data
+   */
+  public deserialize(input) {
+    // this.name = input.name;
```
Comment (cppwfs) [inline ui/src/app/analytics/model/dashboard-item.model.ts]: remove commented code.

Comment (jvalkeal) [inline ui/src/app/analytics/model/dashboard-item.model.ts]: Unused input param / pointless function?

File: ui/src/app/analytics/model/dashboard-item.model.ts [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+import { Subscription } from 'rxjs/Subscription';
+
+/**
+ * Counter model object for the Analytics module.
+ *
+ * @author Gunnar Hillert
+ */
+export class DashboardItem implements Serializable<DashboardItem> {
+
+  public id: number;
```
Comment (cppwfs) [inline ui/src/app/analytics/model/dashboard-item.model.ts]: probably want to use setters and getters

File: ui/src/app/analytics/model/metric-type.model.ts [hunk 25]
```diff
@@ -0,0 +1,29 @@
+import { Serializable } from '../../shared/model';
```
Comment (cppwfs) [inline ui/src/app/analytics/model/metric-type.model.ts]: needs a spec.

File: ui/src/app/analytics/analytics.service.ts [hunk 26]
```diff
@@ -138,4 +161,119 @@ export class AnalyticsService {
...
+
+  removeDashboardItem(index: number): void {
+    if (!this.dashboardItems || this.dashboardItems.length === 0) {
+      return;
+    }
+    this.dashboardItems.splice(index, 1);
+  }
+
+  getAllDashboardItems(): Observable<DashboardItem> {
+    this.addNewDashboardItem();
```
Comment (cppwfs) [inline ui/src/app/analytics/analytics.service.ts]: this line causes new dashboard items to be added when leaving and entering the page.

File: ui/src/app/analytics/model/dashboard-item.model.ts (hunk 24 above)
Comment (cppwfs) [reply ui/src/app/analytics/model/dashboard-item.model.ts]: will open an issue for this.

File: ui/src/app/analytics/model/metric-type.model.ts (hunk 25 above)
Comment (cppwfs) [reply ui/src/app/analytics/model/metric-type.model.ts]: will open an issue for this.

**Comment by cppwfs** [comment]
LGTM will open issues on identified bugs.

**Comment by cppwfs** [comment]
Addressed code review comments, rebased, squashed merged.

**Pull request closed without merging**